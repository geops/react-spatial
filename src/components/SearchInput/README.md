#

This demonstrates the use of SearchInput.

```jsx
import React from  'react'
import SearchInput from 'react-spatial/components/SearchInput';

class SearchInputExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
    };
  }

  render() {
    const { value } = this.state;
    return (
      <div className="tm-search-input-example">
        <SearchInput
          value={value}
          button="Go!!"
          placeholder="Type  some text ..."
          onChange={(e, val) => this.setState({ value: val })}
        />
        <br />
        <div>{`Value searched : ${value}`}</div>
      </div>
    );
  }
}

<SearchInputExample />;
```
