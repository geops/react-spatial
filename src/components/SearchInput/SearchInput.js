import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { MdClose } from 'react-icons/md';
import Button from '../Button';

const propTypes = {
  /**
   * Value of the input
   */
  value: PropTypes.string,

  /**
   * CSS class of the container (input+button).
   */
  className: PropTypes.string,

  /**
   * CSS class of the clear button .
   */
  classNameClearBt: PropTypes.string,

  /**
   * CSS class of the search button.
   */
  classNameSearchBt: PropTypes.string,

  /**
   * Placeholder of the input.
   */
  placeholder: PropTypes.string,

  /**
   * Function launched when the input has the focus.
   */
  onFocus: PropTypes.func,

  /**
   * Function launched when the input looses the focus.
   */
  onBlurInput: PropTypes.func,

  /**
   * Function triggered when the search button is clicked.
   * element.
   */
  onClickSearchButton: PropTypes.func,

  /**
   * Function launched when the user press a key on the input and the list.
   */
  onKeyPress: PropTypes.func,

  /**
   * Function launched when the user change the content of the input.
   */
  onChange: PropTypes.func,

  /**
   * Content for submit button.
   */
  button: PropTypes.any,

  /**
   * Title for the clear button.
   */
  titleClearBt: PropTypes.string,

  /**
   * Title for the search button.
   */
  titleSearchBt: PropTypes.string,

  /**
   * Title for the input text.
   */
  titleSearchInput: PropTypes.string,
};

const defaultProps = {
  button: 'search',
  value: '',
  className: 'tm-search-input',
  classNameClearBt: 'tm-button tm-bt-clear',
  classNameSearchBt: 'tm-button tm-bt-search',
  placeholder: '',
  titleClearBt: '',
  titleSearchBt: '',
  titleSearchInput: '',
  onFocus: () => {},
  onBlurInput: () => {},
  onClickSearchButton: () => {},
  onKeyPress: () => {},
  onChange: () => {},
};

/**
 * This component displays a search input
 */
class SearchInput extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      focus: false,
    };
    this.refInput = null;
  }

  componentDidUpdate(prevProps, prevState) {
    const { focus } = this.state;
    if (this.refInput && focus && focus !== prevState.focus) {
      this.refInput.focus();
    }
  }

  onFocus(evt) {
    this.setState({ focus: true });
    const { onFocus } = this.props;
    onFocus(evt);
  }

  onBlur(evt) {
    this.setState({ focus: false });
    const { onBlurInput } = this.props;
    onBlurInput(evt.nativeEvent, this.lastKeyPress);
  }

  onKeyUp(evt) {
    if (evt.which === 13) {
      this.search(evt);
    }
    const { onKeyPress } = this.props;
    onKeyPress(evt);
    this.lastKeyPress = null;
  }

  onKeyDown(evt) {
    this.lastKeyPress = evt.nativeEvent;
  }

  /**
   * Trigger the onChange function with a new value or the current value of
   * the input.
   */
  search(evt, newVal) {
    const { value, onChange } = this.props;
    if (newVal !== undefined && newVal !== null) {
      onChange(evt, newVal);
    } else {
      onChange(evt, value);
    }
  }

  render() {
    const {
      button,
      className,
      classNameClearBt,
      classNameSearchBt,
      value,
      placeholder,
      titleClearBt,
      titleSearchBt,
      titleSearchInput,
      onClickSearchButton,
    } = this.props;
    const { focus } = this.state;

    // Hide clear button
    const display = !value ? 'none' : undefined;

    let newClassName = className || '';
    if (focus) {
      newClassName = `${newClassName} tm-focus`;
    }

    return (
      <div className={newClassName}>
        <input
          value={value}
          type="text"
          tabIndex="0"
          placeholder={placeholder}
          title={titleSearchInput}
          onChange={e => this.search(e, e.target.value)}
          onBlur={e => this.onBlur(e)}
          onFocus={e => this.onFocus(e)}
          onKeyDown={e => this.onKeyDown(e)}
          onKeyUp={e => this.onKeyUp(e)}
          ref={node => {
            this.refInput = node;
          }}
        />
        <Button
          className={classNameClearBt}
          title={titleClearBt}
          style={
            display && {
              display,
            }
          }
          onClick={e => {
            this.search(e, '');
            this.setState({ focus: true });
          }}
        >
          <MdClose focusable={false} />
        </Button>
        <Button
          className={classNameSearchBt}
          title={titleSearchBt}
          onClick={e => {
            this.search(e);
            onClickSearchButton(e);
          }}
        >
          {button}
        </Button>
      </div>
    );
  }
}

SearchInput.propTypes = propTypes;
SearchInput.defaultProps = defaultProps;

export default SearchInput;
