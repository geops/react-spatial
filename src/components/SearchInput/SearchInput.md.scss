.tm-search-input-example {
  .tm-search-input {
    position: relative;
    width: 300px;
    height: 40px;
    transition: width 0.3s ease;

    &.tm-focus {
      width: 600px;
    }
  }
}
