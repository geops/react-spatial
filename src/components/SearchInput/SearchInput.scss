.tm-search-input {
  display: flex;
  justify-content: flex-end;
  width: calc(100% - 4px);
  height: calc(100% - 4px);
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto;
  border: 1px solid $gray;
  background-color: white;

  input,
  button,
  svg {
    border: none;
  }

  input {
    padding: 0 0 0 15px;
    flex-grow: 2;
    text-overflow: ellipsis;
    min-width: 0;

    &::-ms-clear {
      display: none;
    }
  }

  div[role='button'] {
    width: $btn-size-base;
    height: 100%;
    flex-shrink: 0;
  }

  div[role='button']:last-child {
    background-color: $btn-primary-color;
    color: white;

    &:hover {
      background-color: $btn-primary-color-hover;
    }
  }
}
