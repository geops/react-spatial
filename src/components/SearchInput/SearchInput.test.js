import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SearchInput from './SearchInput';

configure({ adapter: new Adapter() });

describe('SearchInput', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('displays 1 error for required property ', () => {
      shallow(<SearchInput />);
      expect(spy).toHaveBeenCalledTimes(0);
    });

    test('matches snapshot', () => {
      const component = renderer.create(<SearchInput />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('calls default onBlurInput, onFocus, onKeyPress without errors.', () => {
      const wrapper = shallow(<SearchInput />);
      wrapper
        .find('input')
        .first()
        .simulate('focus', {})
        .simulate('blur', {})
        .simulate('keydown', {})
        .simulate('keyup', {});
    });
  });

  describe('when properties are set', () => {
    test('matches snapshot ', () => {
      const component = renderer.create(
        <SearchInput
          value="bar"
          className="tm-foo"
          placeholder="gux"
          onBlur={() => {}}
          onKeyPress={() => {}}
          onChange={() => {}}
        />,
      );
      expect(component.getInstance().state.focus).toBe(false);
      expect(component.toJSON()).toMatchSnapshot();
      component.getInstance().setState({ focus: true });
      expect(component.getInstance().state.focus).toBe(true);
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('matches snapshot when className is undefined ', () => {
      const component = renderer.create(<SearchInput className={undefined} />);
      expect(component.getInstance().state.focus).toBe(false);
      component.getInstance().setState({ focus: true });
      expect(component.getInstance().state.focus).toBe(true);
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('calls onBlurInput, onFocus, onKeyPress properties.', () => {
      const fn = jest.fn();
      const wrapper = shallow(
        <SearchInput onBlurInput={fn} onFocus={fn} onKeyPress={fn} />,
      );
      wrapper
        .find('input')
        .first()
        .simulate('focus', {})
        .simulate('blur', {})
        .simulate('keydown', {})
        .simulate('keyup', {});
      expect(fn).toHaveBeenCalledTimes(3);
    });

    test('searchs on input change event.', () => {
      const wrapper = shallow(<SearchInput />);
      const spy = jest.spyOn(wrapper.instance(), 'search');
      const evt = {
        target: {
          value: 'foo',
        },
      };
      wrapper
        .find('input')
        .first()
        .simulate('change', evt);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(evt, 'foo');
    });

    test('searchs with an empty string on click on cross button.', () => {
      const wrapper = shallow(<SearchInput />);
      const spy = jest.spyOn(wrapper.instance(), 'search');
      const evt = {
        target: {
          value: 'foo',
        },
      };
      expect(wrapper.state('focus')).toBe(false);
      wrapper
        .find('Button')
        .first()
        .simulate('click', evt);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(evt, '');
      expect(wrapper.state('focus')).toBe(true);
    });

    test('searchs on click on search button.', () => {
      const fn = jest.fn();
      const wrapper = shallow(<SearchInput onClickSearchButton={fn} />);
      const spy = jest.spyOn(wrapper.instance(), 'search');
      const evt = {
        target: {
          value: 'foo',
        },
      };
      wrapper
        .find('Button')
        .at(1)
        .simulate('click', evt);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(evt);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test('searchs on enter key up.', () => {
      const wrapper = shallow(<SearchInput />);
      const spy = jest.spyOn(wrapper.instance(), 'search');
      const evt = {
        which: 13,
      };
      wrapper.find('input').simulate('keyup', evt);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(evt);
    });

    test('focuses the input on state change.', () => {
      const wrapper = mount(<SearchInput />);
      expect(document.activeElement).toBe(document.body);
      expect(wrapper.state('focus')).toBe(false);
      wrapper.setState({
        focus: true,
      });
      wrapper.update();
      expect(wrapper.instance().refInput).toBeDefined();
      expect(document.activeElement).toBe(wrapper.instance().refInput);
    });
  });
});
