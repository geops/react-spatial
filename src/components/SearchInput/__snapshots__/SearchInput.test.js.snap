// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput when no properties are set matches snapshot 1`] = `
<div
  className="tm-search-input"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder=""
    tabIndex="0"
    title=""
    type="text"
    value=""
  />
  <div
    aria-label=""
    className="tm-button tm-bt-clear"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "display": "none",
      }
    }
    title=""
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  <div
    aria-label=""
    className="tm-button tm-bt-search"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot  1`] = `
<div
  className="tm-foo"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="gux"
    tabIndex="0"
    title=""
    type="text"
    value="bar"
  />
  <div
    aria-label=""
    className="tm-button tm-bt-clear"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  <div
    aria-label=""
    className="tm-button tm-bt-search"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot  2`] = `
<div
  className="tm-foo tm-focus"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="gux"
    tabIndex="0"
    title=""
    type="text"
    value="bar"
  />
  <div
    aria-label=""
    className="tm-button tm-bt-clear"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  <div
    aria-label=""
    className="tm-button tm-bt-search"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot when className is undefined  1`] = `
<div
  className="tm-search-input tm-focus"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder=""
    tabIndex="0"
    title=""
    type="text"
    value=""
  />
  <div
    aria-label=""
    className="tm-button tm-bt-clear"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "display": "none",
      }
    }
    title=""
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  <div
    aria-label=""
    className="tm-button tm-bt-search"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    title=""
  >
    search
  </div>
</div>
`;
